### PR TITLE
Add WebP support

### PR DESCRIPTION
**Pull request by dhorytskyi**
## Problem

Speed up a page load using another image format with smaller size.

## Solution

Convert and resize loaded images to the WebP format. Send the WebP image if the browser supports it.

## Requested Reviewers

@DrewML

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/architecture?pullRequest=151) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/architecture?pullRequest=151) before we can accept your contribution.<br/><hr/>**Dmytro Horytskyi** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

File: design-documents/webp-support.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+### Design
+
+On image resizing for each size, Magento can create one more image in WebP format.
+For prevention unexpected resource consumption it can be managed by configuration and by default be disabled.
+A web server will be responsible for the decision which file send.
+This decision will be made by availability the requested image with WebP extension and supporting WebP by the browser using the Accept header.
+
+#### Component Dependencies
+
+This functionality requires an asynchronous image resizing to avoid degradation on adding images to the gallery and resizing all images.
```
Comment (roma-glushko) [inline design-documents/webp-support.md]: How will the asynchronous image resizing be implemented? Is it going to be a separated async service that would be triggered by webserver to handle image requests?

Comment (dhorytskyi) [reply design-documents/webp-support.md]: Asynchronous image resizing hasn't been implemented yet.
But in my opinion, when a resized image does not exist, it should be resized in a synchronous way. Asynchronous resizing should be used for adding an image to a media gallery and for all images resizing.

**Comment by DrewML** [comment]
Can we update the proposal to clarify whether this is for catalog images, wysiwyg uploads, or both? My understanding is that these are types are handled separately currently.

**Comment by dhorytskyi** [comment]
I've updated it.
This proposal for catalog images, but it can be introduced for wysiwyg images as well.

**Comment by DrewML** [comment]
We can probably only safely enable this for jpgs. [Lossless png tends to end up larger as a lossless webp image](https://developers.google.com/speed/webp/faq#can_a_webp_image_grow_larger_than_its_source_image), and we can't assume a store owner would want to convert from lossless >> lossy (png >> lossy webp).

**Comment by dhorytskyi** [comment]
As it will be managed by config option, a store owner can decide does it make his images smaller with the same quality or not.

**Comment by michaelcasey316** [comment]
> We can probably only safely enable this for jpgs. [Lossless png tends to end up larger as a lossless webp image](https://developers.google.com/speed/webp/faq#can_a_webp_image_grow_larger_than_its_source_image)

I think the link only says that lossless PNG to lossy WebP (or any lossy to lossless) can increase file size. The issue with file size increasing on a conversion to lossy is because of YUV420 downsampling. Lossless WebP doesn't downsample colors, and near the bottom it states 
> Note that converting a JPEG source to lossy WebP, or a PNG source to lossless WebP are not prone to such file size surprises.

**Comment by DrewML** [comment]
@michaelcasey316 Good catch!

@dhorytskyi do you have any numbers to share for how much savings we get from this on a stock theme with sample data?

**Comment by dhorytskyi** [comment]
I've done a simple test on Magento with sample data using Chrome DevTools.

A category page with **jpeg** images:
**33** / 222 requests | **249 KB** / 1.1 MB transferred | **239 KB** / 3.5 MB resources
jpeg_category

A category page with **webp** images:
**33** / 222 requests | **155 KB** / 1.0 MB transferred | **145 KB** / 3.4 MB resources
webp_category

A product page with **jpeg** images:
**13** / 224 requests | **140 KB** / 234 KB transferred | **157 KB** / 3.5 MB resources
jpeg_product

A product page with **webp** images:
**13** / 224 requests | **90.7 KB** / 184 KB transferred | **107 KB** / 3.4 MB resources
webp_product

**Pull request merged**